### PR TITLE
On TerminatorKind::DropAndReplace still handle unused_mut correctly

### LINKED ISSUE #61424: "variable does not need to be mutable" not detected
(sorry if this is an duplicate or the wrong repo :innocent: )

```rust
fn main() {
    let mut name;
    name = format!("foo{}", "bar");
    dbg!(name);
}
```
([Playground](https://play.integer32.com/?version=stable&mode=debug&edition=2018&gist=50169a334a1863bd3aaed60d83222ba6))

Current behavior:
no warning

Expected behavior:
```
warning: variable does not need to be mutable
 --> src/main.rs:2:9
  |
2 |      let mut name;
  |          ----^
  |          |
  |          help: remove this `mut`
  |
  = note: #[warn(unused_mut)] on by default
```

The variable clearly doesn't need to be mutable, still, nothing is detected and no warning is emitted.

Rust 1.35.0
(Using the [playground](https://play.integer32.com/?version=nightly&mode=debug&edition=2018&gist=0d88ecdb9fa002a3b81c7e7fc48572ab), also in `1.37.0-nightly` (`2019-05-31 7840a0b753a065a41999`))

### PR DESCRIPTION
**Pull request by czipperz**
Closes #61424

- [x] Todo add regression test

### CODE REVIEW

**Comment by czipperz** [comment]
Awaiting the inevitable compiler error because I don't have the toolchain set up locally...

File: src/librustc_mir/borrow_check/used_muts.rs [hunk 1]
```diff
@@ -52,6 +52,24 @@ struct GatherUsedMutsVisitor<'visit, 'cx: 'visit, 'gcx: 'tcx, 'tcx: 'cx> {
     mbcx: &'visit mut MirBorrowckCtxt<'cx, 'gcx, 'tcx>,
 }
 
+impl<'visit, 'cx, 'gcx, 'tcx> GatherUsedMutsVisitor<'visit, 'cx, 'gcx, 'tcx> {
```
Comment (Centril) [inline src/librustc_mir/borrow_check/used_muts.rs]:
```suggestion
impl GatherUsedMutsVisitor<'_, '_, '_, '_> {
```

Comment (czipperz) [reply src/librustc_mir/borrow_check/used_muts.rs]: Shoot I didn't know we could do that.  That's great!

File: src/librustc_mir/borrow_check/used_muts.rs [hunk 2]
```diff
@@ -52,6 +52,24 @@ struct GatherUsedMutsVisitor<'visit, 'cx: 'visit, 'gcx: 'tcx, 'tcx: 'cx> {
     mbcx: &'visit mut MirBorrowckCtxt<'cx, 'gcx, 'tcx>,
 }
 
+impl GatherUsedMutsVisitor<'_, '_, '_, '_> {
+    fn remove_never_initialized_mut_locals(into: &Place) {
```
Comment (Centril) [inline src/librustc_mir/borrow_check/used_muts.rs]: This ain't a method since you have no `self` receiver... :)

**Comment by czipperz** [comment]
When the rustc tutorial says `./x.py check` was fast, they weren't lying!

File: src/librustc_mir/borrow_check/used_muts.rs (hunk 2 above)
Comment (czipperz) [reply src/librustc_mir/borrow_check/used_muts.rs]: Haha!

**Comment by czipperz** [comment]
@matthewjasper It seems that this didn't resolve it.  What should I do?

File: src/test/ui/nll/issue-61424.rs [hunk 3]
```diff
@@ -0,0 +1,7 @@
+#![deny(unused_mut)]
+
+fn main() {
+    let mut x; //~ ERROR: variable does not need to be mutable
+    x = 0;
```
Comment (matthewjasper) [inline src/test/ui/nll/issue-61424.rs]: This needs to be a type that needs drop, e.g., `String`.

**Comment by matthewjasper** [comment]
You can run `./x.py test src/test/ui --stage 1 --bless` to automatically generate the required stderr file.

File: src/test/ui/nll/issue-61424.rs [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+    dbg!(0);
```
Comment (jonas-schievink) [inline src/test/ui/nll/issue-61424.rs]:
```suggestion
    dbg!(x);
```

**Comment by czipperz** [comment]
Well that was pretty fun!  Thanks guys.

**Pull request merged**